server/grpc/runner/runner.go

```diff
@@ -0,0 +1,283 @@
+		case <-ctx.Done():
+			return ctx.Err()
+		case <-stream.Context().Done():
+			return stream.Context().Err()
+		}
+	}
+}
+
+// Result emits the result of an Execution.
+func (s *Server) Result(ctx context.Context, req *ResultRequest) (*ResultResponse, error) {
+	// check authorization and get runner hash
+	runnerHash, err := s.authorize(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// make sure runner is allowed to update this execution
+	var exec *execution.Execution
+	route := fmt.Sprintf("custom/%s/%s/%s", executionmodule.QuerierRoute, executionmodule.QueryGet, req.ExecutionHash)
+	if err := s.rpc.QueryJSON(route, nil, &exec); err != nil {
+		return nil, err
+	}
+	if !exec.ExecutorHash.Equal(runnerHash) {
+		return nil, fmt.Errorf("this runner (%q) is not authorized to submit the result of this execution, the executor should be %q", runnerHash, exec.ExecutorHash)
+	}
+
+	// update execution
+	acc, err := s.rpc.GetAccount()
+	if err != nil {
+		return nil, err
+	}
+	msg := executionmodule.MsgUpdate{
+		Executor: acc.GetAddress(),
+		Hash:     req.ExecutionHash,
+	}
+	switch result := req.Result.(type) {
+	case *ResultRequest_Outputs:
+		msg.Result = &executionmodule.MsgUpdateOutputs{
+			Outputs: result.Outputs,
+		}
+	case *ResultRequest_Error:
+		msg.Result = &executionmodule.MsgUpdateError{
+			Error: result.Error,
+		}
+	}
+	if _, err := s.rpc.BuildAndBroadcastMsg(msg); err != nil {
+		return nil, err
+	}
+	return &ResultResponse{}, nil
+}
+
+// Event emits an event.
+func (s *Server) Event(ctx context.Context, req *EventRequest) (*EventResponse, error) {
+	// check authorization and get runner hash
+	runnerHash, err := s.authorize(ctx)
+	if err != nil {
+		return nil, err
+	}
+
+	// get runner to access instance hash
+	var run *runner.Runner
+	route := fmt.Sprintf("custom/%s/%s/%s", runnermodule.QuerierRoute, runnermodule.QueryGet, runnerHash)
+	if err := s.rpc.QueryJSON(route, nil, &run); err != nil {
+		return nil, err
+	}
+
+	// publish event
+	if _, err := s.ep.Publish(run.InstanceHash, req.Key, req.Data); err != nil {
+		return nil, err
+	}
+
+	return &EventResponse{}, nil
+}
+
+// --------------------------------------------------------
+// Token credential
+// --------------------------------------------------------
+
+// TokenCredential is a structure that manage a token.
+type TokenCredential struct {
+	token string
+}
+
+// NewTokenCredential return a token credential struct that implements credentials.PerRPCCredentials interface.
+func NewTokenCredential(token string) *TokenCredential {
+	return &TokenCredential{
+		token: token,
+	}
+}
+
+// GetRequestMetadata returns the metadata for the request.
+func (c *TokenCredential) GetRequestMetadata(context.Context, ...string) (map[string]string, error) {
+	return map[string]string{
+		"token": c.token,
+	}, nil
+}
+
+// RequireTransportSecurity tells if the transport should be secured.
+func (c *TokenCredential) RequireTransportSecurity() bool {
+	return false
+}
```

Comment:
Not sure this should be here, it is not used by the engine, I would just duplicate it in the e2e test until we have a library the does it for us.

Reply:
I moved this to the services